Migrated project to TypeScript

Anastasia1403 opened the pull request:
replaced all files to .ts and .tsx
changed structure of project
added description to README

Anastasia1403 committed:
Migrated project to TypeScript

replaced all files to .ts and .tsx
changed structure of project
added description to README

Kather1ne commented on src/components/ButtonSubmit/ButtonSubmit.tsx:
```diff
@@ -0,0 +1,16 @@
+  children: React.ReactNode;
+  type?:'button' | 'submit' | 'reset' | undefined
+}
+const ButtonSubmit = function ({ children, type }: ButtonSubmitProps) {
```
можно поставить тип FC<Props>, где FC предполагает наличие children и его не нужно будет прописывать в пропсах

Kather1ne commented on src/components/ButtonSubmit/ButtonSubmit.tsx:
```diff
@@ -0,0 +1,16 @@
+const ButtonSubmit = function ({ children, type }: ButtonSubmitProps) {
+  return (
+    <StyledButton data-testid="submit" type={type}>
+      {children}
```
а что планируется прокидывать сюда как children?  может стоит сделать более читаемым и если там будет иконка или текст, то передавать их пропсами

Anastasia1403 replied:
Тут либо текст, либо текст и иконка, либо иконка и текст

Kather1ne replied:
сейчас не критично, но обычно такие штуки можно передавать пропсами, чтобы контент был понятен

Kather1ne commented on src/components/Card/AppointmentCard.tsx:
```diff
@@ -16,8 +17,9 @@ import {
 } from './styled';
 
-const AppointmentCard = function ({ listItem }) {
+const AppointmentCard = function (listItem: IAppointmentForPatient) {
```
интерфейсы, которые служат для пропсов компонентов нормально держать в самой компоненте, сделав его экспортируемым, и экспортить из компонента в тест. если интерфейсы используются в нескольких компонентах, то их можно вынести в отдельный файл, но тоже не в кучу, а в логически отдельный файл, т.е то что например относится к авторизации будет в одном, а что к доктору в другом файле

Kather1ne commented on src/components/CustomSelect/CustomSelect.tsx:
```diff
@@ -0,0 +1,41 @@
+      {...restProps}
+      value={defaultValue(options, value)}
+  // eslint-disable-next-line
+      onChange={(value: any) => {
+        onChange(value);
+      }}
```
думаю если напишешь так, то не придется дизейблить линтер 
```suggestion
      onChange={onChange}
```

Anastasia1403 replied:
У меня тогда возникают сложности с типизацией этой функции

Kather1ne replied:
тогда возможно стоит поменять типы у аргумента той функции что ты прокидываешь, чтобы они соответствовали параметрам библиотеки. если посмотреть на то что ожидает библиотека, то можно увидеть что в колбеке изменения есть два параметра, из которых нас интересует первый `newValue: SingleValue<string | OptionsType>`. это как раз и можно установить типом для твоего колбэка

Kather1ne commented on src/components/Input/InputStyle.ts:
```diff
@@ -1,10 +1,16 @@
 
+interface StyledInputProps {
+  name: string,
+  error: string | undefined,
```
знак вопроса заменяет `undefined`
```suggestion
  error?: string,
```

Kather1ne commented on src/components/Input/ShowPasswordIcon/ShowPasswordIcon.tsx:
```diff
@@ -1,30 +1,21 @@
-    top: 16px;
-  }
-`;
+type ShowPasswordIconProps = {
```
лучше тут использовать interface.  но если решила для пропсов использовать type, то стоит тогда везде так делать

Kather1ne commented on src/components/UserInfo/UserInfo.tsx:
```diff
@@ -0,0 +1,26 @@
+
+interface UserInfoProps {
+  name: string,
+  userRole: Role | undefined,
```
тут тоже `?`

Kather1ne commented on src/pages/NewAppointment/NewAppointment.tsx:
```diff
@@ -1,33 +1,31 @@
 
 const NewAppointment = function () {
-  const profile = useSelector(userProfile);
+  const profile = useAppSelector(userProfile);
```
почему решила поменять?

Anastasia1403 replied:
Я так понимаю, что это чтобы каждый раз не определять тип стора
Но мне все равно пришлось, т.к. селекторы вынесены отдельно

Kather1ne commented on src/pages/NewAppointment/NewAppointment.tsx:
```diff
@@ -1,33 +1,31 @@
         <StyledNav>
-          <StyledLink to="/user-view/appointments">Doctors</StyledLink>
-          <ArrowIcon alt="arrow" />
+          <StyledLink to={USER_PATH.APPOINTMENTS}>Doctors</StyledLink>
```
можно сделать `enum`

Kather1ne commented on src/pages/NewAppointment/components/NewAppointmentForm/NewAppointmentForm.tsx:
```diff
@@ -1,29 +1,31 @@
 
-  const handleSubmit = (values) => {
-    const formattedValues = {
+  const handleSubmit = (values: any): void => {
```
почему `values: any`?

Kather1ne commented on src/pages/NewAppointment/components/NewAppointmentForm/NewAppointmentForm.tsx:
```diff
@@ -50,25 +52,25 @@ const NewAppointmentForm = function () {
+          <StyledNewAppointmentForm>
             <section>
-              <Stage>1. Select a doctor and define the reason of your visit</Stage>
+              <Stage>{DICT.STAGE.ONE}</Stage>
```
лучше разделить эту большую переменную на несколько енамов. где один будет отвечать за тайтлы, второй за плейсхолдеры. тогда это будет понятнее и читаемее

Kather1ne commented on src/pages/RestorePassword/dictionary.ts:
```diff
@@ -0,0 +1,19 @@
+export const DICT = {
```
повтор

Kather1ne commented on src/pages/UserView/UserView.tsx:
```diff
@@ -0,0 +1,51 @@
+import React, { useEffect } from 'react';
+// import { useHistory } from 'react-router-dom';
```
весь неиспользуемый код убираем

Kather1ne commented on src/pages/UserView/components/AppointmentsList/types.ts:
```diff
@@ -0,0 +1,33 @@
+export interface IPatientCard {
+  id: string,
+  patient: Patient,
+  status: 'canceled' | 'confirmed' | 'waiting',
```
можно вынести в enum

Kather1ne commented on src/pages/UserView/components/AppointmentsList/types.ts:
```diff
@@ -0,0 +1,33 @@
+
+export interface Patient {
+  photo: string,
+  first_name: string,
```
в названии переменных придерживаемся camelCase

Anastasia1403 replied:
это данные, которые с апи приходят ведь

Kather1ne commented on src/pages/NewAppointment/components/TableTime/TableTime.tsx:
```diff
@@ -20,21 +25,22 @@ const TableTime = function ({ blocked }) {
+            htmlFor={label}
             checked={value === values.time}
-            disabled={!time.find((item) => item.value === value)}
+            disabled={!time.find((item: any) => item.value === value)}
```
тут нет необходимости в any

Kather1ne commented on src/pages/SignUp/components/SignUpForm/SignUpForm.tsx:
```diff
@@ -0,0 +1,107 @@
+    password: '',
+    confirmPassword: '',
+  };
+  const handleSubmit = (values: any) => {
```
убрать any

Kather1ne commented on src/redux/addNewAppointment/thunk.ts:
```diff
@@ -1,14 +1,15 @@
       return await instance.post(url.addNewAppointment, appointmentData, { headers: authHeader() })
         .then((response) => response.statusText);
-    } catch (error) {
+    } catch (error: any) {
```
any ставить в совсем крайних случаях, если ты знаешь что тебе приходит в ошибке, то пишем к ней типы

Anastasia1403 replied:
здесь мне приходит объект ошибки, мне важно только то, что лежит в response, а что там остальное мне неважно
как тут типизировать тогда?

Kather1ne replied:
можно обозначить именно то значение, которое ты используешь
`type ResponseError = {error: {response: {data: string }}}`

Kather1ne commented on src/redux/getPatients/selectors.ts:
```diff
@@ -0,0 +1,13 @@
+import { RootState } from 'redux/store';
+import formatDate from '../../utils/formatDate';
+
+export const patientsList = (state: RootState) => state.patients.patients
```
сейчас не критично, но в будущем для понятности того что это селектор можно в конце переменной добавить это обозначение
```suggestion
export const patientsListSelector = (state: RootState) => state.patients.patients
```

Kather1ne commented on src/redux/getSpecializations/slice.ts:
```diff
@@ -0,0 +1,42 @@
+};
+
+const specializationsSlice = createSlice({
+  name: 'getSpecializations',
```
имена в слайсах это не названия для действий, а название твоего стора так что используем существительные тут

Kather1ne commented on src/redux/registration/slice.ts:
```diff
@@ -0,0 +1,33 @@
+    error: '',
+  },
+  reducers: {
+    // registrationSuccess(state, action: PayloadAction<string>) {
```
просмотри пожалуйста везде комментарии

Kather1ne reviewed (commented):
1. посмотри пожалуйста в тестах TableTime у меня ругается на handleChange

2. при обновлении страницы сбрасывает на экран сайнапа

3. при входе через sign in, видимо с неверным паролем
![image](https://user-images.githubusercontent.com/26750258/146228550-46c95bd5-9f71-4091-93c1-f6de2ac9500a.png)

Anastasia1403 committed:
fixed according to PR

fixed according to PR
added drop down menu for card component
added notifications
added Not Found page
added redux-persist

Kather1ne commented on src/components/Card/PatientCard.tsx:
```diff
@@ -20,9 +22,19 @@ import Status from '../Status/Status';
   const { patient, status, visit_date: visitDate } = listItem;
+  const [isDropDownVisible, setIsDropDownVisible] = useState(false);
 
+  const OnChangeVisible = () => {
```
названия функций с маленькой буквы

Kather1ne commented on src/components/DropDown/DropDown.tsx:
```diff
@@ -0,0 +1,17 @@
+  isVisible: boolean
+}
+
+const DropDown = function ({ children, isVisible }: DropDownProps) {
```
опять же, используя тип `FC<DropDownProps>` не нужно будет указывать тип чилдрена как any. 
а так на самом деле не вижу смысла передавать тут чилдрены. можно элементы списка прописать в этой компоненте, раз он используется только в одном месте

Kather1ne commented on src/components/DropDownButton/DropDownButton.tsx:
```diff
@@ -0,0 +1,15 @@
+import styled from 'styled-components';
+
+export const DropDownButton = styled.button`
```
это тоже можно внести в компонент `DropDown`. и для выпадающего меню лучше импользовать элемент ul>li, а не набор кнопок

Kather1ne commented on src/components/EmptyBlock/EmptyBlock.tsx:
```diff
@@ -0,0 +1,26 @@
+  text2?: string,
+}
+
+const EmptyBlock = function ({ text1, text2 }: EmptyBlockProps) {
```
в чем назначение этого компонента?

Anastasia1403 replied:
Когда нет аппоинтментов у пациента или у доктора, будет рисоваться этот компонент

Kather1ne commented on src/pages/NewAppointment/components/NewAppointmentForm/NewAppointmentForm.tsx:
```diff
@@ -54,13 +55,13 @@ const NewAppointmentForm = function () {
           <StyledNewAppointmentForm>
             <section>
-              <Stage>{DICT.STAGE.ONE}</Stage>
+              <Stage>{STAGE.ONE}</Stage>
```
не критично, но само название енамов пишется с первой заглавной, а потом все маленькие. а уже его параметры можно все заглавными обозначать

Kather1ne commented on src/pages/NewAppointment/components/TableTime/TableTime.test.tsx:
```diff
@@ -7,17 +7,48 @@ import { Provider } from 'react-redux';
       },
-      handleChange: (event: React.ChangeEvent<any>): string => mockHandleChange(event.target.value),
+      handleChange: (event: React.ChangeEvent<any>): void => mockHandleChange(event.target.value),
+      touched: true,
+      isSubmitting: false,
+      isValidating: false,
+      errors: false,
+      submitCount: 0,
+      setStatus: jest.fn(),
+      setErrors: jest.fn(),
+      setSubmitting: jest.fn(),
+      setTouched: jest.fn(),
+      setValues: jest.fn(),
+      setFieldValue: jest.fn(),
+      setFieldError: jest.fn(),
+      setFieldTouched: jest.fn(),
+      validateForm: jest.fn(),
+      validateField: jest.fn(),
+      resetForm: jest.fn(),
+      submitForm: jest.fn(),
+      setFormikState: jest.fn(),
+      handleSubmit: jest.fn(),
+      handleReset: jest.fn(),
+      handleBlur: jest.fn(),
+      getFieldProps: jest.fn(),
+      getFieldMeta: jest.fn(),
+      getFieldHelpers: jest.fn(),
+      dirty: false,
+      isValid: true,
+      initialValues: {},
+      initialErrors: {},
+      initialTouched: {},
+      registerField: jest.fn(),
+      unregisterField: jest.fn(),
```
а зачем добавлены все эти параметры?

Anastasia1403 replied:
Тут, как я понимаю, раз я замокала формик контекст, то должна прописать реализацию всех его свойств и методов, иначе у меня не получилось справиться с тем, что ts здесь ругается на типы

Kather1ne commented on src/pages/UserView/components/AppointmentsList/AppointmentsList.tsx:
```diff
@@ -19,15 +21,45 @@ const AppointmentsList = function () {
   );
 };
 
+// <CardList>
```
убрала комментарии в одном месте, но надобавляла в другом)

Kather1ne commented on src/pages/UserView/dictionary.ts:
```diff
@@ -0,0 +1,9 @@
+export const USER_VIEW_DICT = {
```
заглавными буквами мы обозначаем только константы которые не являются объектами (строки, числа, булевые значения). когда объект camelCase, как и его параметры

Kather1ne commented on src/routes/tabsInfo.tsx:
```diff
@@ -19,23 +20,23 @@ export const userTabsInfo: Array<TabInfo> = [
     component: <LoadingBlock />,
     exact: false,
-
+    default: false,
```
за что отвечает этот параметр?

Anastasia1403 replied:
за то,  на какую вкладку пользователя перекидывает по умолчанию, когда он заходит

Kather1ne commented on package.json:
```diff
@@ -25,6 +25,7 @@
     "react-select": "^5.2.1",
     "react-select-event": "^5.3.0",
     "redux": "^4.1.2",
+    "redux-persist": "^6.0.0",
```
зачем добавлена библиотека?

Anastasia1403 replied:
чтобы сохранить данные профиля и статус логинизации между обновлениями страницы и каждый раз не выкидывало на авторизацию

Anastasia1403 committed:
Fix according PR comments

Anastasia1403 committed:
fix imports and add EmptyBlock to PatientsList
